In [81]:
!pip install pandas numpy plotly matplotlib textblob wikipedia requests yfinance python-dotenv
!python -m textblob.download_corpora

[nltk_data] Downloading package brown to /root/nltk_data...
[nltk_data]   Package brown is already up-to-date!
[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package averaged_perceptron_tagger_eng to
[nltk_data]     /root/nltk_data...
[nltk_data]   Package averaged_perceptron_tagger_eng is already up-to-
[nltk_data]       date!
[nltk_data] Downloading package conll2000 to /root/nltk_data...
[nltk_data]   Package conll2000 is already up-to-date!
[nltk_data] Downloading package movie_reviews to /root/nltk_data...
[nltk_data]   Package movie_reviews is already up-to-date!
Finished.


In [82]:
import os
import json
import random
import requests
import wikipedia
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import plotly.express as px
import plotly.graph_objects as go

from textblob import TextBlob
from datetime import datetime, timedelta
import yfinance as yf

# Create folders
os.makedirs("data", exist_ok=True)
os.makedirs("outputs", exist_ok=True)

# Put your API keys here (OPTIONAL)
NEWSAPI_KEY = ""            # Example: "abc123xyz"
ALPHAVANTAGE_KEY = ""

# Companies list
COMPANIES = [
    {"ticker": "AAPL", "name": "Apple Inc"},
    {"ticker": "MSFT", "name": "Microsoft"},
    {"ticker": "GOOGL", "name": "Alphabet Inc"},
]


In [83]:
def fetch_wikipedia_summary(company_name, sentences=4):
    try:
        summary = wikipedia.summary(company_name, sentences=sentences)
        page = wikipedia.page(company_name)
        return {
            "name": company_name,
            "summary": summary,
            "url": page.url
        }
    except:
        return {
            "name": company_name,
            "summary": "Summary not available",
            "url": None
        }

In [84]:
def simulate_google_trends(days=90, seed=42):
    random.seed(seed)
    np.random.seed(seed)

    dates = [datetime.today() - timedelta(days=i) for i in range(days)]
    dates = dates[::-1]  # oldest → newest

    values = np.cumsum(np.random.randn(days))  # random walk
    values = values - values.min()
    values = (values / values.max()) * 100

    df = pd.DataFrame({
        "date": dates,
        "trend_score": values.round(2)
    })
    return df

In [85]:
def fetch_news(company_name):
    if NEWSAPI_KEY == "":
        # Mock sample articles
        return [
            {"title": f"{company_name} announces new upgrade"},
            {"title": f"Analysts give positive feedback on {company_name}"},
            {"title": f"{company_name} faces privacy concerns"},
        ]

    url = "https://newsapi.org/v2/everything"
    params = {
        "q": company_name,
        "pageSize": 20,
        "language": "en",
        "apiKey": NEWSAPI_KEY
    }

    response = requests.get(url, params=params).json()
    return response.get("articles", [])

In [86]:
def classify_sentiment(text):
    polarity = TextBlob(text).sentiment.polarity

    if polarity > 0.1:
        label = "Positive"
    elif polarity < -0.1:
        label = "Negative"
    else:
        label = "Neutral"

    return polarity, label

In [87]:
def fetch_stock(ticker):
    # If Alpha Vantage key missing → use yfinance
    if ALPHAVANTAGE_KEY == "":
        df_yf = yf.download(ticker, period="3mo", interval="1d")

        # Ensure the DataFrame is not empty and has 'Close' column
        if df_yf.empty or 'Close' not in df_yf.columns:
            return pd.DataFrame(columns=["date", "close"])

        # Create a new DataFrame for processing to ensure column names and types are clean
        processed_df = pd.DataFrame({
            "date": df_yf.index, # Use the index (which is 'Date') as the date column
            "close": pd.to_numeric(df_yf['Close'].squeeze(), errors='coerce') # Ensure 'Close' is a Series using .squeeze()
        })

        # Drop any rows where 'close' conversion resulted in NaN
        processed_df = processed_df.dropna(subset=['close'])

        # Reset index to make 'date' a regular column if it's still an index
        # (In this case, it was created from index, so it's already a column, but safe to ensure)
        processed_df = processed_df.reset_index(drop=True)

        return processed_df

    # Alpha Vantage
    url = "https://www.alphavantage.co/query"
    params = {
        "function": "TIME_SERIES_DAILY",
        "symbol": ticker,
        "apikey": ALPHAVANTAGE_KEY
    }

    r = requests.get(url, params=params).json()
    data = r.get("Time Series (Daily)", {})

    if not data:
        return pd.DataFrame(columns=["date", "close"])

    df = pd.DataFrame([
        {"date": k, "close": float(v["4. close"]) } for k, v in data.items()
    ])
    df["date"] = pd.to_datetime(df["date"])
    df = df.sort_values("date")

    return df

In [88]:
import requests
import json
from datetime import datetime

class SlackNotifier:
    """Simple class to send notifications to Slack from Google Colab"""

    def __init__(self, webhook_url):
        """
        Initialize the SlackNotifier with your webhook URL

        Args:
            webhook_url: Your Slack Incoming Webhook URL
        """
        self.webhook_url = webhook_url

    def send_message(self, message, username="Colab Bot", emoji=":robot_face:"):
        """
        Send a simple text message to Slack

        Args:
            message: The text message to send
            username: Bot username that appears in Slack
            emoji: Emoji icon for the bot

        Returns:
            bool: True if successful, False otherwise
        """
        payload = {
            "text": message,
            "username": username,
            "icon_emoji": emoji
        }

        try:
            response = requests.post(
                self.webhook_url,
                data=json.dumps(payload),
                headers={'Content-Type': 'application/json'}
            )

            if response.status_code == 200:
                print(f"✅ Message sent successfully!")
                return True
            else:
                print(f"❌ Failed to send message. Status code: {response.status_code}")
                return False

        except Exception as e:
            print(f"❌ Error sending message: {str(e)}")
            return False

    def send_formatted_message(self, title, message, color="good", fields=None):
        """
        Send a formatted message with attachments to Slack

        Args:
            title: Title of the message
            message: Main message text
            color: Color of the sidebar (good/warning/danger or hex color)
            fields: List of field dictionaries for additional info

        Returns:
            bool: True if successful, False otherwise
        """
        attachment = {
            "title": title,
            "text": message,
            "color": color,
            "footer": "Google Colab",
            "ts": int(datetime.now().timestamp())
        }

        if fields:
            attachment["fields"] = fields

        payload = {
            "attachments": [attachment]
        }

        try:
            response = requests.post(
                self.webhook_url,
                data=json.dumps(payload),
                headers={'Content-Type': 'application/json'}
            )

            if response.status_code == 200:
                print(f"✅ Formatted message sent successfully!")
                return True
            else:
                print(f"❌ Failed to send message. Status code: {response.status_code}")
                return False

        except Exception as e:
            print(f"❌ Error sending message: {str(e)}")
            return False

    def notify_training_complete(self, model_name, accuracy, loss, duration):
        """
        Send a notification when model training is complete

        Args:
            model_name: Name of the model
            accuracy: Final accuracy
            loss: Final loss
            duration: Training duration in seconds
        """
        fields = [
            {"title": "Model", "value": model_name, "short": True},
            {"title": "Accuracy", "value": f"{accuracy:.4f}", "short": True},
            {"title": "Loss", "value": f"{loss:.4f}", "short": True},
            {"title": "Duration", "value": f"{duration:.2f} seconds", "short": True}
        ]

        self.send_formatted_message(
            title="🎉 Training Complete!",
            message=f"Your model '{model_name}' has finished training.",
            color="good",
            fields=fields
        )

    def notify_error(self, error_message, context=""):
        """
        Send an error notification to Slack

        Args:
            error_message: The error message
            context: Additional context about where the error occurred
        """
        fields = [
            {"title": "Error", "value": error_message, "short": False}
        ]

        if context:
            fields.append({"title": "Context", "value": context, "short": False})

        self.send_formatted_message(
            title="⚠️ Error in Colab",
            message="An error occurred during execution",
            color="danger",
            fields=fields
        )

# ============================================
# USAGE EXAMPLES
# ============================================

# Step 1: Set up your Slack Webhook URL
# Go to: https://api.slack.com/messaging/webhooks
# Create an app and add Incoming Webhooks, then copy your webhook URL

# Replace this with your actual webhook URL
WEBHOOK_URL = "https://hooks.slack.com/services/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"

# Initialize the notifier
notifier = SlackNotifier(WEBHOOK_URL)

# Example 1: Send a simple message
notifier.send_message("Hello from Google Colab! 🚀")

# Example 2: Send a formatted message
notifier.send_formatted_message(
    title="Experiment Update",
    message="Your deep learning experiment is running",
    color="warning",
    fields=[
        {"title": "Dataset", "value": "MNIST", "short": True},
        {"title": "Epochs", "value": "50", "short": True}
    ]
)

# Example 3: Notify when training is complete
# This would typically be called after your model training loop
notifier.notify_training_complete(
    model_name="ResNet50",
    accuracy=0.9543,
    loss=0.1234,
    duration=3600.5
)

# Example 4: Send error notifications
try:
    # Your code here
    result = 1 / 0  # This will cause an error
except Exception as e:
    notifier.notify_error(
        error_message=str(e),
        context="Division operation in data preprocessing"
    )

# ============================================
# SIMPLE FUNCTION VERSION (if you prefer)
# ============================================

def send_slack_notification(webhook_url, message):
    """
    Simple function to send a message to Slack

    Args:
        webhook_url: Your Slack webhook URL
        message: Message to send

    Returns:
        bool: True if successful
    """
    payload = {"text": message}

    try:
        response = requests.post(
            webhook_url,
            data=json.dumps(payload),
            headers={'Content-Type': 'application/json'}
        )
        return response.status_code == 200
    except:
        return False

# Usage
# send_slack_notification(WEBHOOK_URL, "Training started! 🏃‍♂️")

✅ Message sent successfully!
✅ Formatted message sent successfully!
✅ Formatted message sent successfully!
✅ Formatted message sent successfully!


In [89]:
wiki_data = []
trend_data = {}
news_data = []
sentiment_data = []
stock_data = {}

for c in COMPANIES:
    name = c["name"]
    ticker = c["ticker"]

    print(f"Processing: {name}")

    # Wikipedia
    wiki_info = fetch_wikipedia_summary(name)
    wiki_data.append({"ticker": ticker, **wiki_info})

    # Simulated Google Trends
    trend_df = simulate_google_trends()
    trend_df["ticker"] = ticker
    trend_data[ticker] = trend_df

    # News & Sentiment
    articles = fetch_news(name)
    for art in articles:
        title = art.get("title", "")
        polarity, label = classify_sentiment(title)

        news_data.append({"ticker": ticker, "title": title})
        sentiment_data.append({"ticker": ticker, "title": title, "polarity": polarity, "label": label})

    # Stock
    stock_df = fetch_stock(ticker)
    stock_df["ticker"] = ticker
    stock_data[ticker] = stock_df

Processing: Apple Inc


/tmp/ipython-input-2076965729.py:4: FutureWarning:

YF.download() has changed argument auto_adjust default to True

[*********************100%***********************]  1 of 1 completed
/tmp/ipython-input-2076965729.py:4: FutureWarning:

YF.download() has changed argument auto_adjust default to True



Processing: Microsoft


[*********************100%***********************]  1 of 1 completed
/tmp/ipython-input-2076965729.py:4: FutureWarning:

YF.download() has changed argument auto_adjust default to True



Processing: Alphabet Inc


[*********************100%***********************]  1 of 1 completed


In [90]:
display(daily_sentiment)

ticker                       date  polarity
0   AAPL 2025-09-05 13:26:51.737173  0.000000
1   AAPL 2025-09-18 13:26:51.737173  0.227273
2   AAPL 2025-10-20 13:26:51.737173  0.136364
3  GOOGL 2025-09-03 13:26:51.737173  0.000000
4  GOOGL 2025-09-18 13:26:51.737173  0.227273
5  GOOGL 2025-10-23 13:26:51.737173  0.136364
6   MSFT 2025-08-22 13:26:51.737173  0.136364
7   MSFT 2025-09-18 13:26:51.737173  0.227273
8   MSFT 2025-11-18 13:26:51.737173  0.000000

In [91]:
wiki_df = pd.DataFrame(wiki_data)
news_df = pd.DataFrame(news_data)
sentiment_df = pd.DataFrame(sentiment_data)

wiki_df.to_csv("data/wiki_summary.csv", index=False)
news_df.to_csv("data/news.csv", index=False)
sentiment_df.to_csv("data/sentiment.csv", index=False)

wiki_df, news_df.head(), sentiment_df.head()

(  ticker          name                                            summary  \
 0   AAPL     Apple Inc  Apple Inc. is an American multinational techno...   
 1   MSFT     Microsoft  Microsoft Corporation is an American multinati...   
 2  GOOGL  Alphabet Inc  Alphabet Inc. is an American multinational tec...   
 
                                            url  
 0     https://en.wikipedia.org/wiki/Apple_Inc.  
 1      https://en.wikipedia.org/wiki/Microsoft  
 2  https://en.wikipedia.org/wiki/Alphabet_Inc.  ,
   ticker                                         title
 0   AAPL               Apple Inc announces new upgrade
 1   AAPL  Analysts give positive feedback on Apple Inc
 2   AAPL              Apple Inc faces privacy concerns
 3   MSFT               Microsoft announces new upgrade
 4   MSFT  Analysts give positive feedback on Microsoft,
   ticker                                         title  polarity     label
 0   AAPL               Apple Inc announces new upgrade  0.136364  Posit

In [92]:
sentiment_summary = sentiment_df.groupby(["ticker", "label"]).size().reset_index(name="count")

fig = px.bar(
    sentiment_summary,
    x="ticker",
    y="count",
    color="label",
    barmode="group",
    title="Sentiment Distribution per Company"
)
fig.show()

In [93]:
for t in trend_data:
    df = trend_data[t]
    fig = px.line(
        df,
        x="date",
        y="trend_score",
        title=f"Simulated Google Trends — {t}"
    )
    fig.show()


In [94]:
for t in stock_data:
    df = stock_data[t]
    fig = px.line(df, x="date", y="close", title=f"Stock Price: {t}")
    fig.show()

# Task
Modify the `fetch_stock` function in cell `53PYriGrjlZy` to pass the 'Close' column directly as a Series to `pd.to_numeric`, then re-run data collection by executing cell `7sTl64iejnld`, recreate DataFrames with cell `lx9S_56ejqEa`, and finally plot the stock prices using cell `m9IG4MEgjvjz`.

## Update fetch_stock function

### Subtask:
Modify the `fetch_stock` function in cell `53PYriGrjlZy` to pass the 'Close' column directly as a Series to `pd.to_numeric` to ensure it's treated as a 1-dimensional input. This is a common and correct usage for `pd.to_numeric` with a pandas Series.


**Reasoning**:
The subtask requires modifying the `fetch_stock` function in cell `53PYriGrjlZy` to correctly pass the 'Close' column as a Series to `pd.to_numeric`. I will replace the existing function with the corrected version.



In [95]:
def fetch_stock(ticker):
    # If Alpha Vantage key missing → use yfinance
    if ALPHAVANTAGE_KEY == "":
        df_yf = yf.download(ticker, period="3mo", interval="1d")

        # Ensure the DataFrame is not empty and has 'Close' column
        if df_yf.empty or 'Close' not in df_yf.columns:
            return pd.DataFrame(columns=["date", "close"])

        # Create a new DataFrame for processing to ensure column names and types are clean
        processed_df = pd.DataFrame({
            "date": df_yf.index, # Use the index (which is 'Date') as the date column
            "close": pd.to_numeric(df_yf['Close'], errors='coerce') # Convert 'Close' to numeric directly
        })

        # Drop any rows where 'close' conversion resulted in NaN
        processed_df = processed_df.dropna(subset=['close'])

        # Reset index to make 'date' a regular column if it's still an index
        # (In this case, it was created from index, so it's already a column, but safe to ensure)
        processed_df = processed_df.reset_index(drop=True)

        return processed_df

    # Alpha Vantage
    url = "https://www.alphavantage.co/query"
    params = {
        "function": "TIME_SERIES_DAILY",
        "symbol": ticker,
        "apikey": ALPHAVANTAGE_KEY
    }

    r = requests.get(url, params=params).json()
    data = r.get("Time Series (Daily)", {})

    if not data:
        return pd.DataFrame(columns=["date", "close"])

    df = pd.DataFrame([
        {"date": k, "close": float(v["4. close"]) } for k, v in data.items()
    ])
    df["date"] = pd.to_datetime(df["date"])
    df = df.sort_values("date")

    return df

**Reasoning**:
Now that the `fetch_stock` function has been updated, the next step in the main task is to re-run the data collection cell `7sTl64iejnld` to apply the changes and re-fetch the data.



In [116]:
wiki_data = []
trend_data = {}
news_data = []
sentiment_data = []
stock_data = {}

for c in COMPANIES:
    name = c["name"]
    ticker = c["ticker"]

    print(f"Processing: {name}")

    # Wikipedia
    wiki_info = fetch_wikipedia_summary(name)
    wiki_data.append({"ticker": ticker, **wiki_info})

    # Simulated Google Trends
    trend_df = simulate_google_trends()
    trend_df["ticker"] = ticker
    trend_data[ticker] = trend_df

    # News & Sentiment
    articles = fetch_news(name)
    for art in articles:
        title = art.get("title", "")
        polarity, label = classify_sentiment(title)

        news_data.append({"ticker": ticker, "title": title})
        sentiment_data.append({"ticker": ticker, "title": title, "polarity": polarity, "label": label})

    # Stock
    stock_df = fetch_stock(ticker)
    stock_df["ticker"] = ticker
    stock_data[ticker] = stock_df

Processing: Apple Inc


/tmp/ipython-input-2076965729.py:4: FutureWarning:

YF.download() has changed argument auto_adjust default to True

[*********************100%***********************]  1 of 1 completed
/tmp/ipython-input-2076965729.py:4: FutureWarning:

YF.download() has changed argument auto_adjust default to True



Processing: Microsoft


[*********************100%***********************]  1 of 1 completed
/tmp/ipython-input-2076965729.py:4: FutureWarning:

YF.download() has changed argument auto_adjust default to True



Processing: Alphabet Inc


[*********************100%***********************]  1 of 1 completed


In [105]:
display(stock_summary_statistics)

count        mean        std         min         25%         50%  \
ticker                                                                     
AAPL     65.0  250.600160  15.456632  224.682190  237.649628  254.183594   
GOOGL    65.0  247.870901  24.754573  199.141891  239.169998  247.139999   
MSFT     65.0  512.381610   9.468442  495.000000  506.739990  510.959991   

               75%         max  
ticker                          
AAPL    262.565491  275.250000  
GOOGL   259.920013  291.309998  
MSFT    517.349976  542.070007

**Reasoning**:
The previous attempt to fix the `TypeError` failed because `df_yf['Close']` was still being interpreted as a multi-dimensional object (likely a DataFrame with one column) by `pd.to_numeric`. To ensure it's treated as a 1-dimensional Series, I need to explicitly apply `.squeeze()` to `df_yf['Close']` before passing it to `pd.to_numeric`. Then, I will re-run the data collection and subsequent cells.



In [106]:
def fetch_stock(ticker):
    # If Alpha Vantage key missing → use yfinance
    if ALPHAVANTAGE_KEY == "":
        df_yf = yf.download(ticker, period="3mo", interval="1d")

        # Ensure the DataFrame is not empty and has 'Close' column
        if df_yf.empty or 'Close' not in df_yf.columns:
            return pd.DataFrame(columns=["date", "close"])

        # Create a new DataFrame for processing to ensure column names and types are clean
        processed_df = pd.DataFrame({
            "date": df_yf.index, # Use the index (which is 'Date') as the date column
            "close": pd.to_numeric(df_yf['Close'].squeeze(), errors='coerce') # Ensure 'Close' is a Series using .squeeze()
        })

        # Drop any rows where 'close' conversion resulted in NaN
        processed_df = processed_df.dropna(subset=['close'])

        # Reset index to make 'date' a regular column if it's still an index
        # (In this case, it was created from index, so it's already a column, but safe to ensure)
        processed_df = processed_df.reset_index(drop=True)

        return processed_df

    # Alpha Vantage
    url = "https://www.alphavantage.co/query"
    params = {
        "function": "TIME_SERIES_DAILY",
        "symbol": ticker,
        "apikey": ALPHAVANTAGE_KEY
    }

    r = requests.get(url, params=params).json()
    data = r.get("Time Series (Daily)", {})

    if not data:
        return pd.DataFrame(columns=["date", "close"])

    df = pd.DataFrame([
        {"date": k, "close": float(v["4. close"]) } for k, v in data.items()
    ])
    df["date"] = pd.to_datetime(df["date"])
    df = df.sort_values("date")

    return df

**Reasoning**:
The `fetch_stock` function has been updated with the `.squeeze()` method. Now, I need to re-run the data collection process by executing cell `7sTl64iejnld` to apply these changes and fetch the stock data correctly.



In [107]:
wiki_data = []
trend_data = {}
news_data = []
sentiment_data = []
stock_data = {}

for c in COMPANIES:
    name = c["name"]
    ticker = c["ticker"]

    print(f"Processing: {name}")

    # Wikipedia
    wiki_info = fetch_wikipedia_summary(name)
    wiki_data.append({"ticker": ticker, **wiki_info})

    # Simulated Google Trends
    trend_df = simulate_google_trends()
    trend_df["ticker"] = ticker
    trend_data[ticker] = trend_df

    # News & Sentiment
    articles = fetch_news(name)
    for art in articles:
        title = art.get("title", "")
        polarity, label = classify_sentiment(title)

        news_data.append({"ticker": ticker, "title": title})
        sentiment_data.append({"ticker": ticker, "title": title, "polarity": polarity, "label": label})

    # Stock
    stock_df = fetch_stock(ticker)
    stock_df["ticker"] = ticker
    stock_data[ticker] = stock_df

Processing: Apple Inc


/tmp/ipython-input-2076965729.py:4: FutureWarning:

YF.download() has changed argument auto_adjust default to True

[*********************100%***********************]  1 of 1 completed
/tmp/ipython-input-2076965729.py:4: FutureWarning:

YF.download() has changed argument auto_adjust default to True



Processing: Microsoft


[*********************100%***********************]  1 of 1 completed
/tmp/ipython-input-2076965729.py:4: FutureWarning:

YF.download() has changed argument auto_adjust default to True



Processing: Alphabet Inc


[*********************100%***********************]  1 of 1 completed


**Reasoning**:
Now that the data collection, including the corrected stock data, has been successfully run, I need to recreate the DataFrames for Wikipedia summaries, news, and sentiment data by executing cell `lx9S_56ejqEa`. This step is crucial for preparing the data for further analysis and visualization as outlined in the main task.



In [108]:
wiki_df = pd.DataFrame(wiki_data)
news_df = pd.DataFrame(news_data)
sentiment_df = pd.DataFrame(sentiment_data)

wiki_df.to_csv("data/wiki_summary.csv", index=False)
news_df.to_csv("data/news.csv", index=False)
sentiment_df.to_csv("data/sentiment.csv", index=False)

wiki_df, news_df.head(), sentiment_df.head()

(  ticker          name                                            summary  \
 0   AAPL     Apple Inc  Apple Inc. is an American multinational techno...   
 1   MSFT     Microsoft  Microsoft Corporation is an American multinati...   
 2  GOOGL  Alphabet Inc  Alphabet Inc. is an American multinational tec...   
 
                                            url  
 0     https://en.wikipedia.org/wiki/Apple_Inc.  
 1      https://en.wikipedia.org/wiki/Microsoft  
 2  https://en.wikipedia.org/wiki/Alphabet_Inc.  ,
   ticker                                         title
 0   AAPL               Apple Inc announces new upgrade
 1   AAPL  Analysts give positive feedback on Apple Inc
 2   AAPL              Apple Inc faces privacy concerns
 3   MSFT               Microsoft announces new upgrade
 4   MSFT  Analysts give positive feedback on Microsoft,
   ticker                                         title  polarity     label
 0   AAPL               Apple Inc announces new upgrade  0.136364  Posit

**Reasoning**:
The data collection and DataFrame creation steps have been successfully completed. The final part of the main task is to plot the stock prices using the data collected. I will now execute cell `m9IG4MEgjvjz` to generate these plots.



In [109]:
for t in stock_data:
    df = stock_data[t]
    fig = px.line(df, x="date", y="close", title=f"Stock Price: {t}")
    fig.show()

## Summary:

### Data Analysis Key Findings
*   The initial attempt to modify the `fetch_stock` function by passing the 'Close' column directly (`df_yf['Close']`) to `pd.to_numeric` resulted in a `TypeError`. This error occurred because, despite `df_yf['Close']` being a pandas Series, `pd.to_numeric` was expecting a strictly 1-dimensional input, which wasn't always guaranteed in that context.
*   The problem was resolved by explicitly ensuring the 'Close' column was a 1-dimensional Series using the `.squeeze()` method (i.e., `df_yf['Close'].squeeze()`). This modification successfully enabled `pd.to_numeric` to process the data correctly.
*   Following the fix, the data collection, DataFrame recreation, and stock price plotting steps executed successfully without errors, confirming the `fetch_stock` function now processes stock data as intended.

### Insights or Next Steps
*   When using `pd.to_numeric` with Series derived from external data sources like `yfinance`, explicitly applying `.squeeze()` can prevent subtle dimensionality errors by ensuring the Series is strictly 1-dimensional.
*   The corrected stock price data is now ready for further analysis, such as calculating moving averages, volatility, or integrating with other data sources like news and sentiment for a more comprehensive stock performance analysis.


# Task
Analyze and visualize the sentiment data by assigning simulated dates, calculating daily average sentiment per company, and plotting the sentiment trends over time.

## Assign Simulated Dates to Sentiment Data

### Subtask:
Add a 'date' column to the `sentiment_df` by assigning random dates within the last 90 days to each sentiment entry.


**Reasoning**:
To assign random dates within the last 90 days to each sentiment entry, I will use `datetime.today()` to get the current date, then generate random days to subtract from it, and finally add these simulated dates as a new 'date' column to the `sentiment_df`.



In [110]:
current_date = datetime.today()
random_days = np.random.randint(0, 90, size=len(sentiment_df))
sentiment_df['date'] = [current_date - timedelta(days=int(d)) for d in random_days]
sentiment_df.head()

ticker                                         title  polarity     label  \
0   AAPL               Apple Inc announces new upgrade  0.136364  Positive   
1   AAPL  Analysts give positive feedback on Apple Inc  0.227273  Positive   
2   AAPL              Apple Inc faces privacy concerns  0.000000   Neutral   
3   MSFT               Microsoft announces new upgrade  0.136364  Positive   
4   MSFT  Analysts give positive feedback on Microsoft  0.227273  Positive   

                        date  
0 2025-10-20 13:47:30.654009  
1 2025-09-18 13:47:30.654009  
2 2025-09-05 13:47:30.654009  
3 2025-08-22 13:47:30.654009  
4 2025-09-18 13:47:30.654009

## Calculate Daily Average Sentiment

### Subtask:
Group the `sentiment_df` by 'ticker' and the newly added 'date' column, then calculate the mean sentiment 'polarity' for each company on each day. This will provide a daily sentiment score.


**Reasoning**:
To calculate the daily average sentiment, I will group the `sentiment_df` by 'ticker' and 'date', then compute the mean of the 'polarity' column and reset the index.



In [111]:
daily_sentiment = sentiment_df.groupby(['ticker', 'date'])['polarity'].mean().reset_index()
daily_sentiment.head()

ticker                       date  polarity
0   AAPL 2025-09-05 13:47:30.654009  0.000000
1   AAPL 2025-09-18 13:47:30.654009  0.227273
2   AAPL 2025-10-20 13:47:30.654009  0.136364
3  GOOGL 2025-09-03 13:47:30.654009  0.000000
4  GOOGL 2025-09-18 13:47:30.654009  0.227273

**Reasoning**:
The daily average sentiment has been calculated. Now, I will visualize the daily sentiment trends for each company using a line plot to show how sentiment polarity changes over time.



In [112]:
fig = px.line(
    daily_sentiment,
    x="date",
    y="polarity",
    color="ticker",
    title="Daily Average Sentiment Trend per Company"
)
fig.show()

## Summary:

### Data Analysis Key Findings

*   A new 'date' column was successfully added to the sentiment data, with each entry assigned a random date within the last 90 days. For example, dates like `2025-10-20` and `2025-09-18` were generated.
*   Daily average sentiment ('polarity') was calculated for each company, grouping the data by 'ticker' and 'date'. This resulted in a DataFrame showing distinct average sentiment scores for companies like 'AAPL' and 'GOOGL' on specific dates.
*   The daily average sentiment trends for each company were visualized using a line plot, effectively showing how sentiment polarity changed over time for different tickers.

### Insights or Next Steps

*   The visualization of daily sentiment trends provides an immediate understanding of how public perception of companies might be fluctuating. This could be useful for identifying periods of strong positive or negative sentiment.
*   Further analysis could involve correlating these sentiment trends with actual stock price movements or significant news events to understand the real-world impact of sentiment.


# Task
Combine the individual stock DataFrames from the `stock_data` dictionary into a single pandas DataFrame.

## Combine Stock Data

### Subtask:
Combine the individual stock DataFrames from the `stock_data` dictionary into a single pandas DataFrame.


**Reasoning**:
To combine all individual stock DataFrames from the `stock_data` dictionary into a single DataFrame, I will use `pd.concat()` and store the result in `all_stock_data`. Then, I will display the first few rows to verify the concatenation.



In [113]:
all_stock_data = pd.concat(stock_data.values(), ignore_index=True)
all_stock_data.head()

date       close ticker
0 2025-08-18  230.666397   AAPL
1 2025-08-19  230.336716   AAPL
2 2025-08-20  225.791107   AAPL
3 2025-08-21  224.682190   AAPL
4 2025-08-22  227.539413   AAPL

**Reasoning**:
The `ValueError: No objects to concatenate` indicates that the `stock_data` dictionary is empty, which means the data collection cell `0dc6fab9` needs to be re-executed to populate it before attempting to concatenate its contents. Assuming `stock_data` will be correctly populated after running `0dc6fab9`, the `pd.concat` code itself is correct for combining the DataFrames.



In [114]:
all_stock_data = pd.concat(stock_data.values(), ignore_index=True)
all_stock_data.head()

date       close ticker
0 2025-08-18  230.666397   AAPL
1 2025-08-19  230.336716   AAPL
2 2025-08-20  225.791107   AAPL
3 2025-08-21  224.682190   AAPL
4 2025-08-22  227.539413   AAPL

**Reasoning**:
The `stock_data` dictionary is currently empty, causing the `ValueError` when `pd.concat` is called. To fix this, I need to re-execute the data collection cell (`0dc6fab9`) first, which populates `stock_data` with the individual stock DataFrames.



In [115]:
wiki_data = []
trend_data = {}
news_data = []
sentiment_data = []
stock_data = {}

for c in COMPANIES:
    name = c["name"]
    ticker = c["ticker"]

    print(f"Processing: {name}")

    # Wikipedia
    wiki_info = fetch_wikipedia_summary(name)
    wiki_data.append({"ticker": ticker, **wiki_info})

    # Simulated Google Trends
    trend_df = simulate_google_trends()
    trend_df["ticker"] = ticker
    trend_data[ticker] = trend_df

    # News & Sentiment
    articles = fetch_news(name)
    for art in articles:
        title = art.get("title", "")
        polarity, label = classify_sentiment(title)

        news_data.append({"ticker": ticker, "title": title})
        sentiment_data.append({"ticker": ticker, "title": title, "polarity": polarity, "label": label})

    # Stock
    stock_df = fetch_stock(ticker)
    stock_df["ticker"] = ticker
    stock_data[ticker] = stock_df

Processing: Apple Inc


/tmp/ipython-input-2076965729.py:4: FutureWarning:

YF.download() has changed argument auto_adjust default to True

[*********************100%***********************]  1 of 1 completed
/tmp/ipython-input-2076965729.py:4: FutureWarning:

YF.download() has changed argument auto_adjust default to True



Processing: Microsoft


[*********************100%***********************]  1 of 1 completed
/tmp/ipython-input-2076965729.py:4: FutureWarning:

YF.download() has changed argument auto_adjust default to True



Processing: Alphabet Inc


[*********************100%***********************]  1 of 1 completed


**Reasoning**:
The traceback indicates that the `fetch_stock` function used during the data collection still lacks the `.squeeze()` method. I need to re-execute the cell (`a68f7d4a`) that defines the corrected `fetch_stock` function to ensure it is applied before re-running the data collection.



In [101]:
def fetch_stock(ticker):
    # If Alpha Vantage key missing → use yfinance
    if ALPHAVANTAGE_KEY == "":
        df_yf = yf.download(ticker, period="3mo", interval="1d")

        # Ensure the DataFrame is not empty and has 'Close' column
        if df_yf.empty or 'Close' not in df_yf.columns:
            return pd.DataFrame(columns=["date", "close"])

        # Create a new DataFrame for processing to ensure column names and types are clean
        processed_df = pd.DataFrame({
            "date": df_yf.index, # Use the index (which is 'Date') as the date column
            "close": pd.to_numeric(df_yf['Close'].squeeze(), errors='coerce') # Ensure 'Close' is a Series using .squeeze()
        })

        # Drop any rows where 'close' conversion resulted in NaN
        processed_df = processed_df.dropna(subset=['close'])

        # Reset index to make 'date' a regular column if it's still an index
        # (In this case, it was created from index, so it's already a column, but safe to ensure)
        processed_df = processed_df.reset_index(drop=True)

        return processed_df

    # Alpha Vantage
    url = "https://www.alphavantage.co/query"
    params = {
        "function": "TIME_SERIES_DAILY",
        "symbol": ticker,
        "apikey": ALPHAVANTAGE_KEY
    }

    r = requests.get(url, params=params).json()
    data = r.get("Time Series (Daily)", {})

    if not data:
        return pd.DataFrame(columns=["date", "close"])

    df = pd.DataFrame([
        {"date": k, "close": float(v["4. close"]) } for k, v in data.items()
    ])
    df["date"] = pd.to_datetime(df["date"])
    df = df.sort_values("date")

    return df

**Reasoning**:
The `fetch_stock` function has been updated with the `.squeeze()` method. Now, I need to re-run the data collection process by executing cell `0dc6fab9` to apply these changes and fetch the stock data correctly.



In [102]:
wiki_data = []
trend_data = {}
news_data = []
sentiment_data = []
stock_data = {}

for c in COMPANIES:
    name = c["name"]
    ticker = c["ticker"]

    print(f"Processing: {name}")

    # Wikipedia
    wiki_info = fetch_wikipedia_summary(name)
    wiki_data.append({"ticker": ticker, **wiki_info})

    # Simulated Google Trends
    trend_df = simulate_google_trends()
    trend_df["ticker"] = ticker
    trend_data[ticker] = trend_df

    # News & Sentiment
    articles = fetch_news(name)
    for art in articles:
        title = art.get("title", "")
        polarity, label = classify_sentiment(title)

        news_data.append({"ticker": ticker, "title": title})
        sentiment_data.append({"ticker": ticker, "title": title, "polarity": polarity, "label": label})

    # Stock
    stock_df = fetch_stock(ticker)
    stock_df["ticker"] = ticker
    stock_data[ticker] = stock_df

Processing: Apple Inc


/tmp/ipython-input-2076965729.py:4: FutureWarning:

YF.download() has changed argument auto_adjust default to True

[*********************100%***********************]  1 of 1 completed


Processing: Microsoft


/tmp/ipython-input-2076965729.py:4: FutureWarning:

YF.download() has changed argument auto_adjust default to True

[*********************100%***********************]  1 of 1 completed


Processing: Alphabet Inc


/tmp/ipython-input-2076965729.py:4: FutureWarning:

YF.download() has changed argument auto_adjust default to True

[*********************100%***********************]  1 of 1 completed


**Reasoning**:
Now that the `stock_data` dictionary has been successfully populated with DataFrames for each company's stock data, I can proceed with combining them into a single DataFrame using `pd.concat()` as per the subtask instructions. This will allow for unified analysis of all stock data.



In [103]:
all_stock_data = pd.concat(stock_data.values(), ignore_index=True)
all_stock_data.head()

date       close ticker
0 2025-08-18  230.666397   AAPL
1 2025-08-19  230.336716   AAPL
2 2025-08-20  225.791107   AAPL
3 2025-08-21  224.682190   AAPL
4 2025-08-22  227.539413   AAPL

## Calculate Summary Statistics

### Subtask:
Calculate descriptive statistics (e.g., mean, standard deviation, min, max) for the 'close' price column, grouped by 'ticker' for the combined stock data.


**Reasoning**:
To calculate the descriptive statistics for the 'close' price column grouped by 'ticker', I will group the `all_stock_data` DataFrame by 'ticker', select the 'close' column, and then use the `.describe()` method.



In [104]:
stock_summary_statistics = all_stock_data.groupby('ticker')['close'].describe()
display(stock_summary_statistics)

count        mean        std         min         25%         50%  \
ticker                                                                     
AAPL     65.0  250.600160  15.456632  224.682190  237.649628  254.183594   
GOOGL    65.0  247.870901  24.754573  199.141891  239.169998  247.139999   
MSFT     65.0  512.381610   9.468442  495.000000  506.739990  510.959991   

               75%         max  
ticker                          
AAPL    262.565491  275.250000  
GOOGL   259.920013  291.309998  
MSFT    517.349976  542.070007

## Final Task

### Subtask:
Summarize the key findings from the stock data summary statistics.


## Summary:

### Q&A
The task was to summarize the key findings from the stock data summary statistics. The key findings include:
*   The `stock_data` dictionary was initially empty, causing a `ValueError` during concatenation.
*   The emptiness of `stock_data` was traced back to a `TypeError` in the `fetch_stock` function, specifically when converting `df_yf['Close']` to numeric, which was resolved by using `.squeeze()` to ensure it was a Series.
*   After the fix, individual stock DataFrames were successfully collected and combined into `all_stock_data`.
*   Descriptive statistics were calculated for the 'close' price, grouped by 'ticker', for AAPL, GOOGL, and MSFT.
*   For AAPL, the mean close price was approximately \$250.60, with a standard deviation of about \$15.46.

### Data Analysis Key Findings
*   The initial attempt to combine individual stock DataFrames failed due to an empty `stock_data` dictionary, which was caused by a `TypeError` in the `fetch_stock` function during the conversion of the 'Close' price column to numeric.
*   The `fetch_stock` function was successfully corrected by adding `.squeeze()` to the 'Close' column before `pd.to_numeric`, ensuring the column was treated as a Series.
*   After the correction, individual stock DataFrames were successfully collected and then combined into a single `all_stock_data` DataFrame.
*   Descriptive statistics for the 'close' price were successfully generated, grouped by 'ticker', providing metrics such as count, mean, standard deviation, and quartiles for AAPL, GOOGL, and MSFT.
*   Specifically, for AAPL, the mean close price was \$250.60, with a standard deviation of \$15.46, indicating its price volatility over the observed period.

### Insights or Next Steps
*   The summary statistics provide a basic understanding of each stock's price behavior; further analysis could involve comparing the volatility (standard deviation) and average prices across the different tickers to identify potential investment opportunities or risks.
*   The data can be used to calculate daily returns, which would then allow for more advanced financial analysis, such as portfolio optimization or risk assessment.
